### ISSUE TITLE
Change model static shape to dynamic shape

### ISSUE THREAD

**Issue by peiwenhuang27** [onnx/onnx#3627]
# Ask a Question

### Question
I have an ONNX model converted from Keras saved model using [tf2onnx](https://github.com/onnx/tensorflow-onnx), which consists of two inputs of static shapes:
> (64, 60, 257) 
> (64, 257, 60, 1) 

I want to change the model shape to dynamic as follows:
> (?, ?, 257)
> (?, 257, ?, 1)

Is there anyway to achieve this using pure ONNX?

### Further information
- Relevant Area (*e.g. model usage, backend, best practices, converters, shape_inference, version_converter, training, test, operators*): I want to use this model in real-time inference where the 1st and 3rd dimensions are both 1 (i.e. shape = [1, 1, 257], [1, 257, 1, 1]), but during training the dimensions are set to a fixed value.

- Is this issue related to a specific model?  
**Model name** (*e.g. mnist*):  [my model link](https://drive.google.com/file/d/1kvUfHFJNYorUcJLVwNKcy1a8bJ1I6pZd/view?usp=sharing)
**Model opset** (*e.g. 7*): 11 

### Notes
I have tried to change the shape with `onnxruntime` like so:
```
# load model into inference session
ONNX_PATH = './model/model.onnx'
model = onnx.load(ONNX_PATH)
# input0: 64, 60, 257 -> N, N, 257
# input1: 64, 257, 60, 1 -> N, 257, N, 1

model.graph.input[0].type.tensor_type.shape.dim[0].dim_param = 'batch_size'
model.graph.input[0].type.tensor_type.shape.dim[1].dim_param = 'seq_len'
model.graph.input[1].type.tensor_type.shape.dim[0].dim_param = 'batch_size'
model.graph.input[1].type.tensor_type.shape.dim[2].dim_param = 'seq_len'

ONNX_PATH = './model/dynamic_model.onnx'
onnx.save(model, ONNX_PATH)
```
Then I tried to inference with the following:
```
so = onnxruntime.SessionOptions()
so.graph_optimization_level = onnxruntime.GraphOptimizationLevel.ORT_DISABLE_ALL
ort_session = onnxruntime.InferenceSession(ONNX_PATH, sess_options=so)
ort_session.set_providers(['CPUExecutionProvider'])

input_name0 = ort_session.get_inputs()[0].name
input_name1 = ort_session.get_inputs()[1].name
output_name = ort_session.get_outputs()[0].name

preds = ort_session.run([output_name], {input_name0: x[:32], input_name1: x_norm[:32]})
```

When running predictions in the last line, the error occurs:
> RuntimeException: [ONNXRuntimeError] : 6 : RUNTIME_EXCEPTION : Non-zero status code returned while running Reshape node. Name:'StatefulPartitionedCall/model/1/Conv2D__36' Status Message: /onnxruntime_src/onnxruntime/core/providers/cpu/tensor/reshape_helper.h:42 onnxruntime::ReshapeHelper::ReshapeHelper(const onnxruntime::TensorShape&, std::vector<long int>&) gsl::narrow_cast<int64_t>(input_shape.Size()) == size was false. The input tensor cannot be reshaped to the requested shape. Input shape:{32,257,60,1}, requested shape:{64,1,257,60}

**Comment by jcwchen** [comment]
Hi @peiwenhuang27, 
Dim can only have either dim_value or dim_param. Thus, besides assigning new dim_param, you need to clear dim_value as well. Something like:
```
model.graph.input[0].type.tensor_type.shape.dim[0].ClearField('dim_value')
model.graph.input[0].type.tensor_type.shape.dim[1].ClearField('dim_value')
model.graph.input[1].type.tensor_type.shape.dim[0].ClearField('dim_value')
model.graph.input[1].type.tensor_type.shape.dim[2].ClearField('dim_value')
```
Since I don't have "x" in your script, I could not test it from my end. Could you please give it a try and see whether the error can be resolved? Thank you!

**Comment by peiwenhuang27** [comment]
Hi @jcwchen, I converted my model from Tensorflow to ONNX with tf2onnx with flag `--inputs input0[-1,-1,257],input1[-1,257,-1,1]` which is supposed to override the model and convert it to dynamic shape.
Then I followed your suggestion and cleared the dim_value and saved the model, but when running it still shows the same error message I attached in the previous comment. I tried to print the model input shapes and they look like this:
```
dim {
  dim_param: "unk__89"
}
dim {
  dim_param: "unk__90"
}
dim {
  dim_value: 257
}

dim {
  dim_param: "unk__91"
}
dim {
  dim_value: 257
}
dim {
  dim_param: "unk__92"
}
dim {
  dim_value: 1
}
```
I guess some ops in the model were not converted successfully, but I'm not sure why.

**Comment by jcwchen** [comment]
Thank you for providing more details and the quick experiment! If you have updated the shape as desired and it still doesn't work, there might be other issues in the converted graph (it makes sense because other corresponding shapes are not updated). I saw you have opened an issue about it under onnx/tensorflow, which is a better place to track it. I think dynamic shape needs to be set before the conversion.

BTW, there is a useful function to update ONNX input/output: https://github.com/onnx/onnx/blob/master/onnx/tools/update_model_dims.py

**Comment by peiwenhuang27** [comment]
As you said, setting the dynamic shape before conversion is critical in my problem. After setting it first before converting the model, the inference works fine, so closing the issue now. Thank you greatly for your suggestion!

**Comment by xingyueye** [comment]
@peiwenhuang27 @jcwchen Hi~ I ran into a similar problem and still blocked in it. Could you please help me.
I need to convert a '.pb' model into '.onnx' using tf2onnx.convert,

> python -m tf2onnx.convert --graphdef model.pb --output model.onnx  --opset 11 --inputs input/image:0[-1,32,128,3],input/seq_len:0[-1] --outputs probablitiesreshape:0,temp_prediction_reshapereshape:0,real_len:0

and then deploy it with 'trtexec'.

> trtexec --onnx=model.onnx --saveEngine=model.engine --explicitBatch --minShapes=input/image:0:1x32x128x3 --optShapes=input/image:0:2x32x128x3 --maxShapes=input/image:0:8x32x128x3 --buildOnly
But Error occurs like that:

> [E] Error[2]: [pointWiseV2Helpers.h::createTensorDesc::296] Error Code 2: Internal Error (Assertion tensor.extent.d[j] == 1 failed.)
Error[2]: [builder.cpp::buildSerializedNetwork::417] Error Code 2: Internal Error (Assertion enginePtr != nullptr failed.)
Segmentation fault (core dumped)

Could you tell me how to fix the onnx, and the input/output node like that:

> name: "input/image:0"
type {
  tensor_type {
    elem_type: 6
    shape {
      dim {
        dim_param: "unk__630"
      }
      dim {
        dim_value: 32
      }
      dim {
        dim_value: 128
      }
      dim {
        dim_value: 3
      }
    }
  }
}

name: "input/seq_len:0"
type {
  tensor_type {
    elem_type: 6
    shape {
      dim {
        dim_param: "unk__631"
      }
    }
  }
}

name: "ctc/probablitiesreshape:0"
type {
  tensor_type {
    elem_type: 1
    shape {
      dim {
        dim_param: "unk__632"
      }
      dim {
        dim_value: 31
      }
    }
  }
}

name: "ctc/temp_prediction_reshapereshape:0"
type {
  tensor_type {
    elem_type: 6
    shape {
      dim {
        dim_param: "unk__633"
      }
      dim {
        dim_value: 31
      }
    }
  }
}

name: "ctc_input/real_len:0"
type {
  tensor_type {
    elem_type: 6
    shape {
      dim {
        dim_param: "unk__634"
      }
    }
  }
}

**Comment by jcwchen** [comment]
Hi @xingyueye,
So if you convert that TensorFlow model to ONNX with dynamic shape, does the conversion also go well? Please try to run `onnx.checker.check_model("your_model.onnx", True)` to validate the converted ONNX model. If the model can pass checker, I would suggest you consult this question with trtexec team. Thanks.

**Comment by xingyueye** [comment]
> Hi @xingyueye, So if you convert that TensorFlow model to ONNX with dynamic shape, does the conversion also go well? Please try to run `onnx.checker.check_model("your_model.onnx", True)` to validate the converted ONNX model. If the model can pass checker, I would suggest you consult this question with trtexec team. Thanks.

Right, This model could successfully pass checker

**Comment by tahiat** [comment]
@[peiwenhuang27](https://github.com/peiwenhuang27), How did you set the dynamic shape before conversion?
I have a pretrained tflite model with input shape (1,1260,960,3) and I want it to be (1,-1,-1,3). I tried to set dynamic shape during conversion by passing the arguments --inputs input_name[1,-1,-1,3] and then cleared the dim_value. but still facing the issue you faced before.
